File: mail-client/src/lib/secretkey.ts
import { readFileSync, appendFileSync, existsSync, mkdirSync, chmodSync } from "fs";
import { randomBytes } from "crypto";

// Mirrors cli/internal/secretkey (Go): keys used to encrypt sensitive values this app
// stores in Postgres (mailbox passwords) live in /etc/patrabahok/secrets.env — the same
// root-only (0600) file the installer and admin dashboard already use — not in the
// database, so a database-only compromise doesn't hand over live mailbox credentials.
const SECRETS_FILE = "/etc/patrabahok/secrets.env";

const cache = new Map<string, Buffer>();

export function ensureSecretKey(name: string): Buffer {
  const cached = cache.get(name);
  if (cached) return cached;

  const existing = readSecret(name);
  if (existing) {
    cache.set(name, existing);
    return existing;
  }

  const raw = randomBytes(32);
  const encoded = raw.toString("base64");

  mkdirSync("/etc/patrabahok", { recursive: true, mode: 0o700 });
  appendFileSync(SECRETS_FILE, `${name}=${encoded}\n`, { mode: 0o600 });
  try {
    chmodSync(SECRETS_FILE, 0o600);
  } catch {
    // best-effort; file was created with 0o600 above already if it didn't exist
  }

  // Re-read rather than trusting our own write, in case a concurrent process won the
  // race and appended its own value first.
  const confirmed = readSecret(name);
  if (!confirmed) {
    throw new Error(`wrote ${name} to ${SECRETS_FILE} but could not read it back`);
  }
  cache.set(name, confirmed);
  return confirmed;
}

function readSecret(name: string): Buffer | null {
  if (!existsSync(SECRETS_FILE)) return null;
  const content = readFileSync(SECRETS_FILE, "utf8");
  const prefix = `${name}=`;
  let last: string | null = null;
  for (const line of content.split("\n")) {
    if (line.startsWith(prefix)) {
      last = line.slice(prefix.length).trim();
    }
  }
  if (!last) return null;
  return Buffer.from(last, "base64");
}

File: cli/internal/secretkey/secretkey.go
// Package secretkey manages small symmetric keys used to encrypt sensitive values this
// application stores in the database (e.g. a connected Cloudflare API token). The key
// itself lives in /etc/patrabahok/secrets.env, the same root-only (0600) file the
// installer already uses for database credentials, so it never needs its own secret
// management story or an installer/migration change to introduce a new one.
package secretkey

import (
	"bufio"
	"crypto/rand"
	"encoding/base64"
	"fmt"
	"os"
	"strings"
)

const SecretsFile = "/etc/patrabahok/secrets.env"

// Ensure returns the 32-byte value stored under name in the secrets file, generating
// and persisting a new random one on first use.
func Ensure(name string) ([]byte, error) {
	if v, ok, err := read(name); err != nil {
		return nil, err
	} else if ok {
		return v, nil
	}

	raw := make([]byte, 32)
	if _, err := rand.Read(raw); err != nil {
		return nil, err
	}
	encoded := base64.StdEncoding.EncodeToString(raw)

	if err := os.MkdirAll("/etc/patrabahok", 0o700); err != nil {
		return nil, err
	}
	f, err := os.OpenFile(SecretsFile, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o600)
	if err != nil {
		return nil, err
	}
	_, writeErr := fmt.Fprintf(f, "%s=%s\n", name, encoded)
	closeErr := f.Close()
	if writeErr != nil {
		return nil, writeErr
	}
	if closeErr != nil {
		return nil, closeErr
	}
	if err := os.Chmod(SecretsFile, 0o600); err != nil {
		return nil, err
	}

	// Re-read rather than trusting our own write, in case a concurrent caller won the
	// race and appended its own value first — the first line for this name that was
	// actually persisted wins for everyone reading afterward.
	v, ok, err := read(name)
	if err != nil {
		return nil, err
	}
	if !ok {
		return nil, fmt.Errorf("wrote %s to %s but could not read it back", name, SecretsFile)
	}
	return v, nil
}

func read(name string) ([]byte, bool, error) {
	f, err := os.Open(SecretsFile)
	if os.IsNotExist(err) {
		return nil, false, nil
	}
	if err != nil {
		return nil, false, err
	}
	defer f.Close()

	prefix := name + "="
	scanner := bufio.NewScanner(f)
	var last string
	for scanner.Scan() {
		line := scanner.Text()
		if strings.HasPrefix(line, prefix) {
			last = strings.TrimPrefix(line, prefix)
		}
	}
	if err := scanner.Err(); err != nil {
		return nil, false, err
	}
	if last == "" {
		return nil, false, nil
	}
	raw, err := base64.StdEncoding.DecodeString(last)
	if err != nil {
		return nil, false, fmt.Errorf("decode %s: %w", name, err)
	}
	return raw, true, nil
}
